SD-555 Document ArgoCD features (#66)
commercetools/telefonistka: docs/argocd.md

# ArgoCD-specific features

While Telefonistka was initially written to be agnostic of the IaC stack some ArgoCD specific features where added recently, this document describes them.

## Commenting diff on PRs

In most cases users directly manipulate Kubernetes manifests in their DRY form (Helm chart/value files or Kustomize configuration), causing a change to have unexpected results in the rendered manifests. Additionally, the state of the in-cluster objects is not always known in advance which prevents the users from knowing the exact change that will happen in the cluster after merging a PR.

> By posting the differences between the cluster objects and the manifests rendered from the PR branch the PR author **and reviewer** can have a better understanding of the effects of a PR merge.

Reviewer: Rephrased for "active voice"
```suggestion
Posting the differences between the cluster objects and the manifests rendered from the PR branch helps the PR author and reviewer better understand the effects of a PR merge.
```

```yaml
argocd:
  allowSyncfromBranchPathRegex: '^env/staging/.*$'
```

> [!NOTE]
> The `ApplicationSet` controller might need to be configured to ignore changes to this specific key, like so:
>
> ```yaml
> spec:
>  ignoreApplicationDifferences:
>    - jsonPointers:
>        - /spec/source/targetRevision
> ```

## AutoMerge "no diff" Promotion PRs

When Telefonistka promote a change it copies the component folder in its entirety. This can lead to situations where a promotion PR is opened but does not affect a promotion target, either because the nature of the change (whitespace/doc) or because the resulting rendered manifests does not change **for the target clusters** (like when you change a target-specific Helm value/Kustomize configuration).

In those cases Telefonistka can auto-merge the promotion PR, saving the effort of merging the PR and preventing future changes from getting an environment drift warning (TODO link).

 This behavior is gated behind the `argocd.autoMergeNoDiffPRs` [configuration key](installation.md).

## Proxy github webhooks

> While not strictly an "ArgoCD feature" Telefonistk ability to proxy webhooks can provide flexability in configuration and securing webhook to ArgoCD server and the AppplicationSet controlelr, see [here](webhook_multiplexing.md)

Author: ```suggestion
While not strictly an "ArgoCD feature" Telefonistka's ability to proxy webhooks can provide greater flexibility in configuration and securing webhook delivery to ArgoCD server and the ApplicationSet controller, see [here](webhook_multiplexing.md)
```
